File: ServerManagement/src/main/java/org/elkoserver/foundation/servermanagement/ServerViaBrokerShutdown.java
package org.elkoserver.foundation.servermanagement;

import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;

import static java.lang.Integer.parseInt;
import static java.nio.charset.StandardCharsets.UTF_8;

public class ServerViaBrokerShutdown {

    public static void main(String[] arguments) throws IOException, InterruptedException {
        if (arguments.length != 4) {
            printUsage();
        } else {
            shutDownServer(arguments[0], parseInt(arguments[1]), arguments[2], arguments[3]);
        }
    }

    private static void printUsage() {
        System.out.println("Usage: BrokerShutdown <host address> <port number> <password> <server name>");
    }

    private static void shutDownServer(String hostAddress, int portNumber, String password, String serverName) throws IOException, InterruptedException {
        try (Socket socket = new Socket(hostAddress, portNumber)) {
            OutputStream outputStream = socket.getOutputStream();
            outputStream.write(("{to:\"admin\", op:\"auth\", auth:{mode:\"password\", code:\"" + password + "\"}}\n\n").getBytes(UTF_8));
            Thread.sleep(1_000L);
            outputStream.write(("{to:\"admin\", op:\"shutdown\", server:\"" + serverName + "\", kill:false}\n\n").getBytes(UTF_8));
            Thread.sleep(1_000L);
        }
    }
}

File: ServerManagement/src/main/kotlin/org/elkoserver/foundation/servermanagement/ServerViaBrokerShutdown.kt
package org.elkoserver.foundation.servermanagement

import net.sourceforge.argparse4j.helper.HelpScreenException
import net.sourceforge.argparse4j.inf.ArgumentParser
import net.sourceforge.argparse4j.inf.ArgumentParserException
import java.net.Socket
import kotlin.system.exitProcess

object ServerViaBrokerShutdown {
    @JvmStatic
    fun main(arguments: Array<String>) {
        val parser = createParser()
        try {
            tryToShutDownServer(parser, arguments)
        } catch (argumentParserException: ArgumentParserException) {
            if (argumentParserException is HelpScreenException) {
                // No action needed. Usage has already been output.
            } else {
                println(argumentParserException.message)
                parser.printUsage()
                exitProcess(1)
            }
        }
    }

    private fun createParser() =
            createHostPortPasswordArgumentParser("ServerViaBrokerShutdown", false).apply {
                addArgument("server")
                        .required(true)
                        .help("Server name")
            }

    private fun tryToShutDownServer(parser: ArgumentParser, arguments: Array<String>) {
        val parsedArguments = parser.parseArgs(arguments)
        val hostAddress = parsedArguments.getString("host") ?: throw IllegalStateException()
        val portNumber = parsedArguments.getInt("port") ?: throw IllegalStateException()
        val password = parsedArguments.getString("password") ?: throw IllegalStateException()
        val serverName = parsedArguments.getString("server") ?: throw IllegalStateException()
        shutDownServer(hostAddress, portNumber, password, serverName)
    }

    private fun shutDownServer(hostAddress: String, portNumber: Int, password: String, serverName: String) {
        Socket(hostAddress, portNumber).use { socket ->
            val outputStream = socket.getOutputStream()
            outputStream.write("""{"to":"admin", "op":"auth", "auth":{"mode":"password", "code":"$password"}}$END_OF_COMMAND""".toUtf8())
            Thread.sleep(1000L)
            outputStream.write("""{"to":"admin", "op":"shutdown", "server":"$serverName", "kill":false}$END_OF_COMMAND""".toUtf8())
            Thread.sleep(1000L)
        }
    }
}
